doc/cbf/cbf_hld.md

```diff
@@ -0,0 +1,215 @@
+- If a packet is matched against an FC value and the route for its destination does not reference a CBF NHG, the packet will use the route's NH
+- If a packet is matched against an FC value and the route for its destination references a CBF NHG which maps the packet's FC value, the packet will use the mapped NHG
+- If a packet is matched against an FC value and the route for its destination references a CBF NHG which doesn't map the packet's FC value, the packet will be dropped
+
+## 2.2 Configuration and Management Requirements
+- DSCP/EXP to FC maps must be allowed to be configured via the 2 CONFIG_DB tables with no requirement to be configurable via CLI.
+
+## 2.3 Scalability Requirements
+- Unchanged.
+
+## 2.4 Warm Boot Requirements
+- Unchanged - the new class based next hop group table must be compatible with existing warm boot requirements.
+
+## 2.5 Restrictions
+- fpmsyncd is not updated to use the new CLASS_BASED_NEXT_HOP_GROUP_TABLE as part of this enhancement. Anyone wishing to use this feature must use a modified version of fpmsyncd, or program the table directly.
+
+# 3 Design
+## 3.1 Overview
+This design directly changes CONFIG_DB, APPL_DB, orchagent and sairedis.
+
+## 3.2 DB Changes
+### 3.2.1 APPL DB
+Based on the next hop group split (https://github.com/Azure/SONiC/pull/712) on which this HLD is based on, a new CLASS_BASED_NEXT_HOP_GROUP table will be added to the APPL_DB with the following format:
+```
+### CLASS_BASED_NEXT_HOP_GROUP_TABLE
+    ;Stores a list of FC-aware next hop groups.
+    ;Status: Mandatory
+    key           = CLASS_BASED_NEXT_HOP_GROUP_TABLE:string ; arbitrary string identifying the class based next hop group, as determined by the programming application.
+    members       = NEXT_HOP_GROUP_TABLE.key,    ; one or more indexes within NEXT_HOP_GROUP_TABLE, separated by “,”
+    class_map     = number:number,   ; one or more mapping from Forwarding Class to index in "members" field to use as NHG, separated by ","
+```
+
+Example:
+    127.0.0.1:6379[1]> hgetall  "CLASS_BASED_NEXT_HOP_GROUP:CbfNhg1"
+    1) "members"
+    2) "Nhg1,Nhg2,Nhg3,Nhg4"
+    3) "class_map"
+    4) "0:0,1:0,2:1,3:1,4:2,5:2,6:3,7:3"
+
+The ROUTE_TABLE is updated to allow the "nexthop_group" to allow both keys from NEXT_HOP_GROUP_TABLE and from the new CLASS_BASED_NEXT_HOP_GROUP_TABLE.
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+    nexthop_group = NEXT_HOP_GROUP_TABLE.key or CLASS_BASED_NEXT_HOP_GROUP_TABLE.key ; index within the NEXT_HOP_GROUP_TABLE or CLASS_BASED_NEXT_HOP_GROUP_TABLE, optionally used instead of nexthop and intf fields
```

Comment:
@TACappleman Could you also add the changes to Sonic orchagent(if any) when some or all members of CBF NHG in ROUTE_TABLE are not resolved?

Reply:
If a `ROUTE_TABLE` entry references a CBF NHG which hasn't been added yet by the user application, the behavior is the same as for a normal NHG - the route entry will remain in the Consumer's queue for the event of the NHG being created at some point which will allow the route entry to be created.

If a `ROUTE_TABLE` entry references a CBF NHG which has been added by the user application, but doesn't exist in the `ASIC_DB` (because the NH objects weren't created yet in the ASIC_DB perhaps), the same thing as in the first case will happen. `RouteOrch` will keep the entry in the queue, trying to create it. Beyond that, the CBF NHG handler will also keep trying to create the CBF NHG. This behavior is covered by the following statement from the HLD:
`otherwise the task will be kept in the process queue for the event of the missing member(s) being created which would allow the class based next hop group to be created.`